pkg/ccl/changefeedccl/kvfeed/kv_feed.go

```diff
@@ -230,14 +234,27 @@ func (f *kvFeed) run(ctx context.Context) (err error) {
 		return nil
 	}
 
-	// highWater represents the point in time at or before which we know
+	// Frontier initialized to initialHighwater timestamp which
+	// represents the point in time at or before which we know
 	// we've seen all events or is the initial starting time of the feed.
-	highWater := f.initialHighWater
+	frontier, err := span.MakeFrontierAt(f.initialHighWater, f.spans...)
+	if err != nil {
+		return err
+	}
+
 	for i := 0; ; i++ {
 		initialScan := i == 0
-		if err = f.scanIfShould(ctx, initialScan, highWater); err != nil {
+		scannedSpans, scannedTS, err := f.scanIfShould(ctx, initialScan, frontier.Frontier())
+		if err != nil {
 			return err
 		}
+		// We have scanned scannedSpans up to and including scannedTS.  Advance frontier
+		// for those spans -- we can start their range feed from scannedTS.Next().
+		for _, sp := range scannedSpans {
+			if _, err := frontier.Forward(sp, scannedTS.Next()); err != nil {
+				return err
+			}
+		}
```

Comment:
This still concerns me because by setting `frontier` like this we're still saying "We've seen everything up to time scannedTS.Next()" due to the definition of `frontier` and then we rely on that definition later on even though it's no longer necessarily true.  This'd likely never happen but suppose a schema change backfill occurred at time `t` when we had an `endTime` of `t.Next()` while we also had a single change occurring at `t.Next()` as well, wouldn't that event be erroneously skipped?

If our `initialScanOnly` check directly below this block was instead an `endTimeReached = f.endTime.EqOrdering(frontier.Frontier())`, that should technically still have been a reasonable choice but this type of invalidation would've broken that.

Reply:
No, I think we are saying that Scan has seen all events  scanTS, and thus we can start rangefeed after that time -- i.e at scanTS.Next()

I think this is all pretty subtle, and maybe @ajwerner can illucidate/correct things.
Schema change occurred at T, runUntilTableEvent exists with T.Prev().
We then scan table up-to, and including T.  We emit all events at T. That means we can start rangefeed from T.Next().

Reply:
Right, but wouldn't the `endTime` case I described still be possible?  Where we start the rangefeed from `T.next()` but we think we've already finished all events for `T.next()` because that's what the Frontier is at so we allow ourselves to exit at the `endTime` `scanBoundary`.

Reply:
Sorry, I'm just seeing that you mean newly added EndTime -- which simply didn't exist when this PR was written.
I think it's fine?  Or at the very least we didn't define end_time too exactly.  So, it looks like it's exclusive.

Reply:
Yeah, we could define it that way in this case, tho it does feel a little awkward that its "almost always inclusive except for a specific edge case".  

I don't disagree that the way things are set up now will likely work perfectly fine for our _current_ requirements, my concern is more on the complexity/code-smell side where we're adding complexity to the definition of a "frontier" where now it has two possible meanings, either "the time at or before we know we've seen all events" _or_ "the starting point of the rangefeeds".  Bugs/restrictions like "`endTime` can't be inclusive" can occur where we assume one meaning while the other meaning is active.

Here is my suggestion:
<img width="750" alt="Screen Shot 2022-04-20 at 2 44 09 PM" src="https://user-images.githubusercontent.com/6236424/164300914-321dc5e2-7d4c-4723-988f-44ecbe9f454d.png">

This way, our frontier _always_ maintains the invariant of "we've handled all information up to and including the frontier", there's no more "or is the initial starting time of the feed".  It is also always updated as early as possible, so:
1. We initialize it to `initialHighWater` because in `Config` we've defined that as "InitialHighWater is the timestamp **after** which new events are guaranteed to be produced", so `initialHighWater` is already considered handled
2. We immediately forward the checkpoint information because we know that progress has happened from the beginning.  I also don't think we necessarily need to clear out `checkpoint` and `checkpointTimestamp` because now unlike before we have the `timestamp` information that restricts its validity.
3. As soon as the scan occurs we forward the frontier to the `scannedTS` because we know that those spans were progressed up to and including that timestamp.  No further explanation needed of "we'll do `.Next()` so that later down in the code when we start the rangefeeds they start correctly".
4. The `SpanTimePair` initialization is more clear where the intent of "We want to start the rangefeeds _right after_ our existing progress" is clearly laid out within the context it has relevance to.